include/commons.hpp

```diff
@@ -30,64 +30,94 @@
-  int ii = x / ms.extent(1); \
-  int ij = x % ms.extent(1);
+#define dim2(x, ms)            \
+    int ii = x / ms.extent(1); \
+    int ij = x % ms.extent(1);
 // get mdspan 3d indices from 1d index
-#define dim3(x, ms)                            \
-  int ii = x / (ms3.extent(1) * ms.extent(2)); \
-  int ij = (x / ms.extent(2)) % ms.extent(1);  \
-  int ik = x % ms.extent(2)
+#define dim3(x, ms)                              \
+    int ii = x / (ms3.extent(1) * ms.extent(2)); \
+    int ij = (x / ms.extent(2)) % ms.extent(1);  \
+    int ik = x % ms.extent(2)
 
 class Timer {
- public:
-  Timer() { start(); }
-
-  ~Timer() { stop(); }
-
-  void start() { start_time_point = std::chrono::high_resolution_clock::now(); }
-
-  double stop() {
-    end_time_point = std::chrono::high_resolution_clock::now();
-    return duration();
-  }
-
-  double duration() {
-    auto start = std::chrono::time_point_cast<std::chrono::microseconds>(
-                     start_time_point)
-                     .time_since_epoch()
-                     .count();
-    auto end =
-        std::chrono::time_point_cast<std::chrono::microseconds>(end_time_point)
-            .time_since_epoch()
-            .count();
-    auto duration = end - start;
-    double ms = duration * 0.001;
-    return ms;
-  }
-
- private:
-  std::chrono::time_point<std::chrono::high_resolution_clock> start_time_point;
-  std::chrono::time_point<std::chrono::high_resolution_clock> end_time_point;
+   public:
+    Timer() { start(); }
+
+    ~Timer() { stop(); }
+
+    void start() { start_time_point = std::chrono::high_resolution_clock::now(); }
+
+    double stop() {
+        end_time_point = std::chrono::high_resolution_clock::now();
+        return duration();
+    }
+
+    double duration() {
+        auto start =
+            std::chrono::time_point_cast<std::chrono::microseconds>(start_time_point).time_since_epoch().count();
+        auto end = std::chrono::time_point_cast<std::chrono::microseconds>(end_time_point).time_since_epoch().count();
+        auto duration = end - start;
+        double ms = duration * 0.001;
+        return ms;
+    }
+
+   private:
+    std::chrono::time_point<std::chrono::high_resolution_clock> start_time_point;
+    std::chrono::time_point<std::chrono::high_resolution_clock> end_time_point;
+};
+
+enum class sch_t { INVALID = -1, CPU, GPU, MULTIGPU };
+
+class heq_sch {
+   public:
+    heq_sch(const std::string& input) noexcept : scheduler(lookupScheduler(input)) {
+        if (scheduler == sch_t::INVALID) {
+            throw std::invalid_argument("Invalid scheduler type provided.");
+        }
+    }
+
+    heq_sch() noexcept : scheduler(sch_t::CPU) {}
+
+    // Use default for trivial constructors
+    heq_sch(const heq_sch& other) noexcept = default;
+
+    [[nodiscard]] sch_t operator()() const noexcept { return scheduler; }
+
+   private:
+    sch_t scheduler;
+
+    static constexpr std::array<std::pair<std::string_view, sch_t>, 3> schmap = {
```

Comment:
for small elements like 3ish, we can use static memory std::array instead of std::map. The difference in lookup time is generally small for small elements.

Reply:
that's fine too but considering the list of schedulers may grow in the future. Using `map` can also remove the need to explicitly use a for loop as we can just use `map::contains(key)`

Reply:
that makes sense, will do, let me edit my code and address this round of comments.